taf/testlib/common/loops.py

```diff
@@ -0,0 +1,303 @@
+            try:
+                status = self.__next__()
+            except StopIteration:
+                break
+
+        return status
+
+
+class CondLoopInterval(CondLoopImpl):
+    """
+    @brief:  Add an interval to wait before proceeding with the next loop
+    """
+    OPTION = 'interval'
+
+    @classmethod
+    def add_interval(cls, cloop, interval, *args, **kwargs):
+        cloop._interval = interval
+
+        def f():
+            if cloop.loops_counter:
+                do_sleep = True
+                with suppress(AttributeError):
+                    do_sleep = cloop.loops_counter < cloop._count_max
+                if do_sleep:
+                    time.sleep(cloop._interval)
+
+        cloop._before_next = f
+
+    add = add_interval
+
+    def __init__(self, interval):
+        super(CondLoopInterval, self).__init__()
+        self.add_interval(self, interval)
+
+    @property
+    def interval(self):
+        return self._interval
+
+
+class CondLoopCounter(CondLoopImpl):
+    """
+    @brief:  Add a loop counter end condition that fails after the amount of loops is reached.
+    """
+    OPTION = 'counter'
+
+    @classmethod
+    def add_counter(cls, cloop, count, *args, **kwargs):
+        cloop._count_max = count
+
+        def end_cond():
+            if not(cloop.loops_counter < cloop._count_max):
+                return True
+            return False
+
+        cloop.end_conds_fail.append(end_cond)
+
+    add = add_counter
+
+    def __init__(self, count):
+        super(CondLoopCounter, self).__init__()
+        self.add_counter(self, count)
+
+    @property
+    def count_max(self):
+        return self._count_max
+
+
+class CondLoopTimeout(CondLoopImpl):
+    """
+    @brief:  Add a timeout end condition that fails after the deadline has been reached.
+    """
+    OPTION = 'timeout'
+
+    @classmethod
+    def add_timeout(cls, cloop, timeout, *args, **kwargs):
+        cloop._start_time = int(time.time())
+        cloop._deadline = cloop._start_time + timeout
+
+        def end_cond():
+            if not(int(time.time()) < cloop._deadline):
+                return True
+            return False
+
+        cloop.end_conds_fail.append(end_cond)
+
+    add = add_timeout
+
+    def __init__(self, timeout):
+        super(CondLoopTimeout, self).__init__()
+        self.add_timeout(self, timeout)
+
+    @property
+    def start_time(self):
+        return self._start_time
+
+    @property
+    def deadline(self):
+        return self._deadline
+
+
+class ConditionalLoop(CondLoopImpl):
+    """
+    public API
+    """
+    BP_UNREG_OPTIONS = "Unregistered options detected({0}): {1}"
+    BP_UNREG_PAIR = "'{0}'='{1}'"
+
+    def __init__(self, **kwargs):
+        super(ConditionalLoop, self).__init__()
+
+        for cl_option, cl_type in self.OPTIONS.items():
+            with suppress(KeyError):
+                cl_value = kwargs.pop(cl_option)
+                cl_type.add(self, cl_value)
+
+        if kwargs:
```

Comment:
Makes perfect sense not to populate OPTIONS before its been validated, thanks. 

Do we drop the OrderedDict for the sets and with it any means of control over the ordering in which the options are processed? Might not have much impact now for the current default options already there, but for potential future extensions, for the sake of keeping it generic?